posthog/management/commands/fix_future_person_created_at.py

```diff
@@ -0,0 +1,70 @@
+from datetime import timedelta
+import logging
+
+import structlog
+from django.core.management.base import BaseCommand
+from django.db import transaction
+from dateutil.parser import isoparse
+
+from posthog.kafka_client.client import KafkaProducer
+from posthog.models.person.person import Person
+from posthog.models.person.util import create_person
+from django.utils.timezone import now
+
+logger = structlog.get_logger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class Command(BaseCommand):
+    help = "Fix update persons created_at to not be in the future for a single team"
+
+    def add_arguments(self, parser):
+        parser.add_argument("--team-id", default=None, type=int, help="Specify a team to fix data for.")
+        parser.add_argument("--new-date", default="2024-01-01", type=str, help="New create_at value to use")
+        parser.add_argument("--live-run", action="store_true", help="Run changes, default is dry-run")
+
+    def handle(self, *args, **options):
+        run(options)
+
+
+def run(options):
+    live_run = options["live_run"]
+
+    if not options["team_id"]:
+        logger.error("You must specify --team-id to run this script")
+        exit(1)
+
+    team_id = options["team_id"]
+    new_date = isoparse(options["new_date"])
+
+    future_date = now() + timedelta(days=2)
+
+    # Get all persons with future created_at value
+    persons = Person.objects.filter(team_id=team_id, created_at__gt=future_date)
+
+    logger.info(
+        f'Found {len(persons)} persons with future created_at value, updating them to {new_date.strftime("%Y-%m-%d %H:%M:%S.%f")}'
+    )
+
+    # If someone else updated the person at the same time these could conflict, which isn't ideal, but this is a one-off script
+    for person in persons:
+        logger.info(f'Updating person {person.uuid} created_at to {new_date.strftime("%Y-%m-%d %H:%M:%S.%f")}')
+        if live_run:
+            with transaction.atomic():
+                person.created_at = new_date
+                person.version += 1
+                person.save()
```

Comment:
I don't think the `atomic` is doing anything here, since you're only touching one row. There's nothing here preventing us from setting `version` to a lower version (if it was updated by plugin-server in the meantime). It will also (IIRC) replace all fields on the Person with whatever is in your object in memory, so we could lose properties.

I'm rusty, but I think you'd need to do something like,

```python
from django.db.models import F
Person.objects.filter(pk=person.id).update(version=F('version') + 1, created_at=new_date)
```

Which should do something like: `UPDATE posthog_person WHERE id = $1 SET version = version + 1, createt_at = $2`

Reply:
🤦‍♀️  on the atomic.
For the race problem: this is better thanks. It might not fully cover all the races iirc plugin-server does:
1. read
2. compute update
3. write

So we can end up in the middle of those two ... annoyingly. That said If I here don't override properties and plugin-server overrides created_at, then I can just re-run the script.